File: python/22. Generate Parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        res = []
        self.dfs(n, n, '', res)
        return res
        
    def dfs(self, l, r, s, res):
        if l > r:
            return
        if l == r and l == 0:
            res.append(s)
        if l > 0:
            self.dfs(l-1, r, s + '(', res)
        if r > 0:
            self.dfs(l, r-1, s + ')', res)

File: go/src/leetcode/22. Generate Parentheses/22. Generate Parentheses.go
package main

func generateParenthesis(n int) []string {
	ans := []string{}
	dfs(n, n, []uint8{}, &ans)
	return ans
}

func dfs(l, r int, arr []uint8, ans *[]string) {
	if l > r {
		return
	}
	if l==r && l==0 {
		*ans = append(*ans, string(arr))
		return
	}

	if l > 0 {
		dfs(l-1, r, append(arr, '('), ans)
	}
	if r > 0 {
		dfs(l, r-1, append(arr, ')'), ans)
	}
}
